### PR TITLE
Feature/bimpm

### PR DESCRIPTION
**Pull request by bwanglzu**
#183 Not done yet.

### CODE REVIEW

**Comment by rgtjf** [comment]
@lixinsu Cool!

The major difference between the implementation and the paper is the `cosine operation`. For paper, it decodes as `m_k = cosine(W_k \odot v_1, W_k \odot v_1)`. However, we need to broadcast to tensors. In other words, we need the  `cosine operation` as below:
```python
def COS(H_P, H_Q; W):  
    """
    Params:
      H_P: (time_steps, d)
      H_Q: (time_steps, d)
      W: (l, d)
    Returns: (time_steps(p), time_steps(q), l)
    """
    pass
```

Maybe the code from the author's repo will help us.
https://github.com/zhiguowang/BiMPM/blob/master/src/match_utils.py#L101-L132

**Comment by uduse** [comment]
Add me to the reviewers once you think this branch's ready.

**Comment by rgtjf** [comment]
@lixinsu @uduse Pls help to review this branch. 
The basic framework is ready but I am worried about it.

File: matchzoo/layers/attention_layer.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        :param att_type: int
+        :param remove_diagnoal: bool
+        """
+        self.att_dim = att_dim
+        self.att_type = att_type
+        self.remove_diagnoal = remove_diagnoal
+        super(AttentionLayer, self).__init__(**kwargs)
+
+    @property
+    def get_att_type(cls):
```
Comment (lixinsu) [inline matchzoo/layers/attention_layer.py]: Why use class property?

File: matchzoo/models/bimpm_model.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        """Build model structure."""
+        # ~ Input Layer
+        input_left, input_right = self._make_inputs()
+
+        # Word Representation Layer
+        # TODO: concatenate word level embedding and character level embedding.
+        embedding = self._make_embedding_layer()
+        embed_left = embedding(input_left)
+        embed_right = embedding(input_right)
+
```
Comment (lixinsu) [inline matchzoo/models/bimpm_model.py]: Add dropout layer after embedding layer according to original implementations

File: matchzoo/layers/attention_layer.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        class: `AttentionLayer` constructor.
+
+        :param att_dim: int
+        :param att_type: int
+        :param remove_diagnoal: bool
+        """
+        self.att_dim = att_dim
+        self.att_type = att_type
+        self.remove_diagnoal = remove_diagnoal
+        super(AttentionLayer, self).__init__(**kwargs)
```
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/layers/attention_layer.py]: > style: use private attr (`self._att_dim`) unless you have a good reason. 

OK.

> Also, call `super` before everything else unless you have a good reason.

I referred to  https://keras.io/layers/writing-your-own-keras-layers.

File: matchzoo/layers/attention_layer.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        >>> import matchzoo as mz
+        >>> layer = mz.layers.AttentionLayer(att_dim=50)
+
+    """
+
+    def __init__(self,
+                 att_dim: int,
+                 att_type: str = 'default',
+                 remove_diagnoal: bool = False,
+                 **kwargs):
```
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/layers/attention_layer.py]: Also, I referred to https://keras.io/layers/writing-your-own-keras-layers/.

File: matchzoo/layers/attention_layer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/layers/attention_layer.py]: Thanks! I've read it.

File: matchzoo/layers/multi_perspective_layer.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+    batch_nums = K.tf.tile(batch_nums, multiples=[1, pair_size])
+
+    # shape (batch_size, pair_size, 2)
+    indices = K.tf.stack((batch_nums, positions), axis=2)
+    pair_probs = K.tf.gather_nd(probs, indices)
+    # pair_probs = tf.reshape(pair_probs, shape=[batch_size, pair_size])
+    return pair_probs
+
+
+def multi_perspective_match(mp_dim, reps_lt, reps_rt,
```
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/layers/multi_perspective_layer.py]: OK.

File: matchzoo/processor_units/processor_units.py [hunk 6]
```diff
@@ -10,9 +10,12 @@
```
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/processor_units/processor_units.py]: OK.

File: matchzoo/models/bimpm_model.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params(with_embedding=True)
+        params['optimizer'] = 'adam'
+
+        params.add(engine.Param('dim_word_embedding', 50))
+        params.add(engine.Param('dim_char_embedding', 50))
+        params.add(engine.Param('word_embedding_mat', None))
+        params.add(engine.Param('char_embedding_mat', None))
+        params.add(engine.Param('embedding_random_scale', 0.2))
```
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/models/bimpm_model.py]: I will remove unused params in the last version.

File: matchzoo/models/bimpm_model.py (hunk 2 above)
Comment (rgtjf) [reply matchzoo/models/bimpm_model.py]: Done. https://github.com/NTMC-Community/MatchZoo/pull/423/files#diff-6ff9ed022c0ccfe040c58be8933584f5R112

File: matchzoo/models/bimpm_model.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+    @classmethod
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params(with_embedding=True)
+        params['optimizer'] = 'adam'
+
+        params.add(engine.Param('dim_word_embedding', 50))
+        params.add(engine.Param('dim_char_embedding', 50))
+        params.add(engine.Param('word_embedding_mat', None))
```
[the remark this answers is not in the document]
Comment (rgtjf) [reply matchzoo/models/bimpm_model.py]: I have removed the `None`

File: matchzoo/layers/attention_layer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (uduse) [reply matchzoo/layers/attention_layer.py]: You're making them `property`, but the examples you referred is not. I read keras source code before I made that comment, and I found nothing that suggests customized arguments should appear before `super`.

File: matchzoo/layers/attention_layer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (uduse) [reply matchzoo/layers/attention_layer.py]: > If you think kwargs is quite helpful and should be kept, document it.

The example you referred is not a part of keras source code, so it does not need to worry about docs.
